### ISSUE TITLE
2.1.4-rc3 - Not Working on Android

### ISSUE THREAD

**Issue by handstandsam** [tomakehurst/wiremock#455]
Client logs:
```
java.lang.NoSuchMethodError: No direct method <init>(Ljava/lang/String;Lorg/apache/http/entity/ContentType;)V in class Lorg/apache/http/entity/StringEntity; or its super classes (declaration of 'org.apache.http.entity.StringEntity' 
```

Location of this issue: https://github.com/tomakehurst/wiremock/blob/2.0-beta/src/main/java/com/github/tomakehurst/wiremock/client/HttpAdminClient.java

Line 232: 
```java
private String postJsonAssertOkAndReturnBody(String url, String json, int expectedStatus) {
        ...
            StringEntity stringEntity = new StringEntity(json, APPLICATION_JSON);
            post.setEntity(stringEntity);
       ...
}
```

That StringEntity constructor is not supported on the Apache HTTP Client on Android.  Please refer to a previous PR which fixed this to allow support on Android - https://github.com/tomakehurst/wiremock/pull/330/files

Usable on Android:
```Java
StringEntity stringEntity = new StringEntity(json);
stringEntity.setContentType(APPLICATION_JSON.getMimeType());
post.setEntity(stringEntity);
```

This should be updated and documented on why this method signature is used.  It looks like this slipped back in as it's the more concise way to write it.  Also, if there are any other usages of this constructor, those should be updated to use this as well.

**Comment by handstandsam** [comment]
I did some more digging.

It looks like someone is getting HttpClient working on a newer version (HttpClient v4.4.1) with Android, but it's packaged with a different package name "cz.msebera.android.httpclient": https://github.com/smarek/httpclient-android

It does seem to have the updated constructor however, so if it can be built under the original package name, it might? work?

**Comment by tomakehurst** [comment]
I switched the ``StringEntity`` setup deliberately as the setter based approach breaks character encoding (and a corresponding test case introduced in another PR).

I tested it on Android using your repo (or Abha's fork to be precise) and the test passed so I assumed that was OK. Abha said it worked fine for her too. Could it be an Android SDK version issue?

I did also take a quick look at the HTTP client lib you've referenced, but as you say it's in different packages so I couldn't just drop it in. If you have any ideas how to cleanly integrate it, I'd be keen to hear them.

**Comment by handstandsam** [comment]
@abhagupta - I just checked out master branch from https://github.com/abhagupta/AndroidHttpMockingExamples...

It compile worked fine, but same as I was experiencing, running the tests fails at runtime.

```
I/TestRunner: started: testWiremock(com.handstandsam.httpmocking.tests.wiremock.WireMockActivityInstrumentationTestCase2)
I/MonitoringInstrumentation: Activities that are still in CREATED to STOPPED: 0
D/MonitoringInstrumentation: execStartActivity(context, ibinder, ibinder, activity, intent, int, bundle
D/LifecycleMonitor: Lifecycle status change: com.joshskeen.weatherview.MainActivity@b90e7b8 in: PRE_ON_CREATE
D/LifecycleMonitor: Lifecycle status change: com.joshskeen.weatherview.MainActivity@b90e7b8 in: CREATED
D/LifecycleMonitor: Lifecycle status change: com.joshskeen.weatherview.MainActivity@b90e7b8 in: STARTED
D/LifecycleMonitor: Lifecycle status change: com.joshskeen.weatherview.MainActivity@b90e7b8 in: RESUMED
D/OpenGLRenderer: Use EGL_SWAP_BEHAVIOR_PRESERVED: true

                  [ 07-06 23:14:37.220  4680: 4680 D/         ]
                  HostConnection::get() New Host Connection established 0x7fbbe28345c0, tid 4680


                  [ 07-06 23:14:37.262  4680: 4704 D/         ]
                  HostConnection::get() New Host Connection established 0x7fbbe2a34380, tid 4704
I/OpenGLRenderer: Initialized EGL, version 1.4
W/art: Long monitor contention event with owner method=void java.nio.SocketChannelImpl$SocketAdapter.<init>(java.net.PlainSocketImpl, java.nio.SocketChannelImpl) from SocketChannelImpl.java:503 waiters=0 for 355ms
I/TestRunner: failed: testWiremock(com.handstandsam.httpmocking.tests.wiremock.WireMockActivityInstrumentationTestCase2)
I/TestRunner: ----- begin exception -----
java.lang.NoSuchMethodError: No direct method <init>(Ljava/lang/String;Lorg/apache/http/entity/ContentType;)V in class Lorg/apache/http/entity/StringEntity; or its super classes (declaration of 'org.apache.http.entity.StringEntity' appears in /system/framework/org.apache.http.legacy.boot.jar)
at com.github.tomakehurst.wiremock.client.HttpAdminClient.postJsonAssertOkAndReturnBody(HttpAdminClient.java:235)
at com.github.tomakehurst.wiremock.client.HttpAdminClient.addStubMapping(HttpAdminClient.java:104)
at com.github.tomakehurst.wiremock.client.WireMock.register(WireMock.java:229)
at com.github.tomakehurst.wiremock.client.WireMock.register(WireMock.java:225)
at com.github.tomakehurst.wiremock.client.WireMock.givenThat(WireMock.java:100)
at com.github.tomakehurst.wiremock.client.WireMock.stubFor(WireMock.java:104)
at com.handstandsam.httpmocking.tests.wiremock.WireMockActivityInstrumentationTestCase2.testWiremock(WireMockActivityInstrumentationTestCase2.java:70)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
at com.github.tomakehurst.wiremock.junit.WireMockRule$1.evaluate(WireMockRule.java:72)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at android.support.test.internal.runner.TestExecutor.execute(TestExecutor.java:54)
at android.support.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:240)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1879)
```

**Comment by handstandsam** [comment]
I just spent 2 hours trying different combinations of libraries to no avail.  I can't spend any more time on this.

**The Problem:** Android blocks you from providing an alternate httpclient implementation version.  That's because they bundled in a crappy version a long time ago and just recently ripped it out in API 23.  I'm using API 23 and trying to see if I can provide my own, but still can't.  There is no way to provide another version of "StringEntity" because of this.

**Best Idea:** Create a "AndroidCompatibleStringEntity" class that extends StringEntity.  Use "AndroidCompatibleStringEntity" instead and copy in the constructor's implementation that is being used only once in the WireMock codebase.  You can super everything else.

**Another Idea**
Using Reflection use [org.apache.http.entity.StringEntityHC4](http://svn.apache.org/repos/asf/httpcomponents/httpclient-android/branches/test-branch/src/main/java/org/apache/http/entity/StringEntityHC4.java) if it's on the classpath instead.  It's part of the android httpclient impl as a workaround for this.  It's essentially the same as the solution above, but is already provided with `org.apache.httpcomponents:httpclient-android:4.3.5.1`.

**Comment by tomakehurst** [comment]
I think you're probably right - the HC4 version might be the safest bet. Guess it'll mean attempting to detect it on the classpath and falling back to the standard version if it's not present.

**Comment by handstandsam** [comment]
@tomakehurst - I don't recommend the reflection thing as it's hacky/ugly... I think the copy of the class "StringEntity" as "AndroidCompatibleStringEntity" or "WireMockStringEntity" is probably the best.

**Comment by abhagupta** [comment]
@handstandsam  Yes, the test `WireMockActivityInstrumentationTestCase2` failed with wiremock-2.1.1-beta.jar but passed with wiremock-2.1.5.rc4.jar . Can you try the release candidate and confirm?

**Comment by handstandsam** [comment]
@abhagupta - I had to build WireMock from the master branch's source to get 2.1.5-rc4.jar, but your project DID work with 2.1.5-rc4.jar.  I'm kinda dumbfounded as I don't understand why it works, but that's great.

@tomakehurst - Is it easy to push up 2.1.5-rc4 to mavenCentral so I can give it a try on my repo with only changing the WireMock version?

So, I'm guessing it maybe had something with using httpclient 4.5.1? - https://github.com/tomakehurst/wiremock/commit/37ba883f553d462a1229c97d51636fbcab401332

**Comment by handstandsam** [comment]
Woohoo!  I took the 2.1.5-rc4 build I made locally and it's pom file, put it on a local maven repo and then ran my project with success!

Thanks @abhagupta and @tomakehurst - I'd say this can be closed once an official 2.1.5-rc4 is pushed and that version can be added to the [AndroidHttpMockingExamples project](https://github.com/handstandsam/AndroidHttpMockingExamples).

Cheers!

**Comment by handstandsam** [comment]
Just because I spent many hours on trying to figure this out...  I'd like to see if you had any idea what made this work between 2.1.1-beta and 2.1.5-rc4?

**Comment by tomakehurst** [comment]
Ah good, glad it works for you too.

I got it to work by switching to a constructor that took the charset as a String. It is deprecated in the current version of HTTPClient, but still present so I guess it had been around since since the 4.0 version Android provides by default.

**Comment by handstandsam** [comment]
Thanks for the explanation, sounds great!  I updated my examples project (https://github.com/handstandsam/AndroidHttpMockingExamples) to use 2.1.5-rc4 and it's working great.  I'm closing this as it's now working 👍
